### PR TITLE
Convert SYCL example code to Latex

### PR DESCRIPTION
**Pull request by Wee-Free-Scot**
Re-format SYCL example code into Latex for side doc

### CODE REVIEW

**Review by rzambre — COMMENTED** [review]
This renders as an Example section that is immediately followed by an Example label.
For example:
```
3.1 Example: MPI plus SYCL
Example 3.1 This SYCL example demonstrates the usage of the different memory allocation kinds to perform communication in a manner that is supported by the underlying
MPI library.
```
Feels redundant. Not sure which one is the better one to get rid of. Was the primary reason to introduce sections for clearer distinctions for the start of the examples?

**Comment by garzaran** [comment]
> This renders as an Example section that is immediately followed by an Example label. For example:
> 
> ```
> 3.1 Example: MPI plus SYCL
> Example 3.1 This SYCL example demonstrates the usage of the different memory allocation kinds to perform communication in a manner that is supported by the underlying
> MPI library.
> ```
> 
> Feels redundant. Not sure which one is the better one to get rid of. Was the primary reason to introduce sections for clearer distinctions for the start of the examples?

How about something like this to make it less repetitive? It still has the memory allocation kinds, but it provides more details of what the program is doing 

"This SYCL example shows how to query the backend used (Level Zero, CUDA, or HIP) and the usage of the different memory allocation kinds supported by the underlaying mpi library. "

**Comment by Wee-Free-Scot** [comment]
> This renders as an Example section that is immediately followed by an Example label. For example:
> 
> ```
> 3.1 Example: MPI plus SYCL
> Example 3.1 This SYCL example demonstrates the usage of the different memory allocation kinds to perform communication in a manner that is supported by the underlying
> MPI library.
> ```
> 
> Feels redundant. Not sure which one is the better one to get rid of. Was the primary reason to introduce sections for clearer distinctions for the start of the examples?

My primary motivation for adding the `subsection` Latex instructions was to get subsection bookmarks into the document structure for easy navigation to an individual example.

Perhaps just removing the word "Example:" from the subsection name would be sufficient to alleviate the repetition?

In the fullness of time (likely in a future update version for this side-doc) it would be nice to have some text in between the subsection header and the start of the Example environment -- text that explains why the example exists and the key features to which the reader should pay particular attention. Thus, eventually this will read "3.1 MPI plus SYCL", "paragraph or two of text". then "Example 3.1: ..."

File: mem_alloc.tex [hunk 1]
```diff
@@ -141,6 +141,368 @@ \chapter{Examples}
 This section includes examples demonstrating the usage of
 memory kinds defined in \chapref{chap:definitions}.
 
+\section{Example: MPI plus SYCL}
```
Comment (Wee-Free-Scot) [inline mem_alloc.tex:144]:
```suggestion
\section{MPI plus SYCL}
```

File: mem_alloc.tex [hunk 2]
```diff
@@ -327,6 +689,8 @@ \chapter{Examples}
 \end{lstlisting}
 \end{example}
 
+\section{Example: MPI plus ROCm}
```
Comment (Wee-Free-Scot) [inline mem_alloc.tex:692]:
```suggestion
\section{MPI plus ROCm}
```

File: mem_alloc.tex [hunk 3]
```diff
@@ -141,6 +141,368 @@ \chapter{Examples}
...
+    catch (sycl::exception const& e) {
+        std::cout << "An exception was caught.\n";
+        std::terminate();
+    }
+    return 0;
+}
+\end{lstlisting}
+\end{example}
+
+\section{Example: MPI plus CUDA}
```
Comment (Wee-Free-Scot) [inline mem_alloc.tex:504]:
```suggestion
\section{MPI plus CUDA}
```

**Comment by rzambre** [comment]
> It still has the memory allocation kinds, but it provides more details of what the program is doing

> In the fullness of time (likely in a future update version for this side-doc) it would be nice to have some text in between the subsection header and the start of the Example environment -- text that explains why the example exists and the key features to which the reader should pay particular attention. Thus, eventually this will read "3.1 MPI plus SYCL", "paragraph or two of text". then "Example 3.1: ..."

Got it. I would anticipate the titles of the section would also reflect the intent of the example rather than the programming model used in the example. For example, "MPI plus ROCm" would change to "Device-aware File I/O" or something more suitable along those lines (cc @jdinan @edgargabriel ).

> Perhaps just removing the word "Example:" from the subsection name would be sufficient to alleviate the repetition

The suggested changes sound good to me.

**Comment by edgargabriel** [comment]
> Got it. I would anticipate the titles of the section would also reflect the intent of the example rather than the programming model used in the example. For example, "MPI plus ROCm" would change to "Device-aware File I/O" or something more suitable along those lines (cc @jdinan @edgargabriel ).

I would be fine with that if we do it for all examples.

**Pull request merged**